### PR TITLE
[table info][2/4] add utils for table info backup and restore and redesign the db read

### PR DESCRIPTION
**Pull request by jillxuu**
### This PR
1. add gcs operator utils

### Overview
**[1/4] separate indexer async v2 db from aptosdb : https://github.com/aptos-labs/aptos-core/pull/11799
[2/4] add gcs operational utils to set up for backup and restore to gcs: https://github.com/aptos-labs/aptos-core/pull/11793
[3/4] add epoch based backup logic: https://github.com/aptos-labs/aptos-core/pull/11794
[4/4] add db snapshot restore logic: https://github.com/aptos-labs/aptos-core/pull/11795**

### Goal
https://www.notion.so/aptoslabs/Internal-Indexer-Deprecation-Next-Steps-824d5af7f16a4ff3aeccc4a4edee2763?pvs=4
1. Migrating internal indexer off of aptos db critical path, move it into its own standalone runtime service.
7. Still provide the table info mapping to both API and indexer services.
8. improve table info parsing performance to unblock indexer perf bottleneck

### Context
This effort is broken down into two parts:
1. part 1: https://github.com/aptos-labs/aptos-core/pull/10783. Add moving the table info service out of the critical path and convert it into multithread to concurrently process the request.
3. part 2 is this PR. Now we have the table info service, but this service will only be enabled by a handful FNs, so when a new FN wants to join the network but they don't want to sync from the genesis, they should be able to restore the db from a cloud service. In order to provide such a cloud service for others to download the db snapshot, this pr focuses on two things: backup and restore. backup is optional in the config, but restore logic is written in the code.

| before | after |
| --- | --- |
| Screenshot 2023-11-15 at 10 15 01 PM | Screenshot 2024-01-25 at 11 14 05 AM |

### Detailed Changes
Screenshot 2024-01-25 at 11 56 53 AM

#### backup based on epoch or transaction version? what frequency?
Pros of backup using version:
1. We have more control over the backup frequency as frequency is tunable. 
Cons of backup using version:
1. Overhead of managing and comparing backed up versions and current processing versions

Pros of backup using epoch:
1. When to backup logic is much cleaner and less error prone
Cons of backup using epoch:
1. not very configurable but still tunable by setting frequency on how many epochs behind

Decided to use epoch, because on testnet we have little over 10k epoch, divided by total txns, it gives us on average 70k txns per epoch, this sounds about the frequency of txns we'd like to backup.

#### when to restore
Decided to restore when both conditions are met: 
1. the difference btw next versions to be processed and current version from ledger is greater than a `version_diff`.
2. and time difference btw last restored timestamp in db and current timestamp is greater than a `RESTORE_TIME_DIFF_SECS`.
This is to prevent fullnode crashlooping and constantly try to restore without luck, and when version difference is not that big we don't need to spam the gcs service but rather directly state syncing from that close to head version.

#### structure of the gcs bucket
I followed the similar structure as of indexer's filestore, where we keep a `metadata.json` file in the bucket to keep track of the chain id and newest backed up epoch. and then a files folder to keep all the epoch based backup. Each db snapshot is first compressed into a tar file from folder, and then gzipped to compress to the best size possible. based on larry's point, alternative compression like bzip2 is less performant.

#### threads
Using a separate thread for backup only, base on the past experience, gcs upload could be as slow as minutes.

#### gcs prunning
Couple options we could pursue, since each backup file is a full db backup, 
1. create another service to constantly clean up the backup files
2. use gcs own policy to delete files based on time, and other conditions
4. programmatically delete old files while upload
9. constantly writing to the same file

Decided to go with gcs own policy with the proper configuration setup in the gcs deployment. Reason behind is that deploying and maintaining another service is overhead and costs more money, especially this service's responsibility is very singular; writing code for gcs object deletion is not ideal, since we're writing and deleting, need to handle different multitude of edge cases; constantly writing to the same file is def not gonna work, since gcs has strict limitation on single object write limit, only once per second.

### Test Plan
1. passing all the written unit tests on file system operation
10. locally tested and verified backup and restore, as well as table info read

### Concerns
There's a bottleneck on the size of db snapshot. Currently this db on testnet is around 250mb, based on the nature of db, the compression could get it to be 50-150mb per db snapshot. It's still too big to upload to gcs as its too slow. 

#### E2E test
from rustie:
1. set up a long-running fullnode in a new k8s project. We can use the same data-staging-us-central1 cluster, but use a new namespace, like indexer-fullnode-testnet-test  or something. This is to isolate it from everything else, but we can use the same cluster for simplicity
11. Set up a job in the same namespace that does the backup to GCS.
12. We can set up a continuous job in aptos-core CI that:
5.1. Spins up a fullnode based on the latest build. This would be the latest main nightly for instance
3.2. We can quit immediately after verifying that the restore was successful
3.3. The cost should be manageable, assuming that the restore process is quick.

#### load test
Couple things i want to verify with load testing
1. 10 fns boostrapping, can restore work for all of them
2. fns keep crashlooping, is gcs spammed based on egress & ingress
3. when file gets bigger, will backup still work and how long it could take

### CODE REVIEW

**Comment by bowenyang007** [comment]
I would say that clean up shouldn't happen right away. Let's onboard the ecosystem before we delete the "old way"

**Comment by jillxuu** [comment]
> I would say that clean up shouldn't happen right away. Let's onboard the ecosystem before we delete the "old way"

old way was not deleted, and it remains exactly the same. we're just changing the way to read and write our new db. @bowenyang007

File: api/types/src/convert.rs [hunk 1]
```diff
@@ -422,19 +429,32 @@ impl<'a, R: ModuleResolver + ?Sized> MoveConverter<'a, R> {
...
+                // If table_info_reader exists, use it to get table_info
+                Ok(table_info_reader.get_table_info(handle)?)
+            } else if self.db.indexer_enabled() {
+                // If indexer is enabled in db, get table_info from the db
+                Ok(Some(self.db.get_table_info(handle)?))
+            } else {
+                // If neither condition is met, return None
+                Ok(None)
+            }
+        };
```
Comment (bowenyang007) [inline api/types/src/convert.rs:444]: I think that you still need the self.db.indexer_async_v2_enabled check no?

File: config/src/config/indexer_table_info_config.rs [hunk 2]
```diff
@@ -4,8 +4,8 @@
 use serde::{Deserialize, Serialize};
 
 // Useful defaults
-pub const DEFAULT_PARSER_TASK_COUNT: u16 = 20;
-pub const DEFAULT_PARSER_BATCH_SIZE: u16 = 1000;
+pub const DEFAULT_PARSER_TASK_COUNT: u16 = 10;
+pub const DEFAULT_PARSER_BATCH_SIZE: u16 = 100;
```
Comment (bowenyang007) [inline config/src/config/indexer_table_info_config.rs:8]: why?

File: api/types/src/convert.rs (hunk 1 above)
Comment (jillxuu) [reply api/types/src/convert.rs:444]: no longer needed, cuz we add indexer async v2's own db reader to the API context, so that it's exposed for all FN APIs, same applies to indexer FN. `aptos-node/src/services.rs`

File: config/src/config/indexer_table_info_config.rs (hunk 2 above)
Comment (jillxuu) [reply config/src/config/indexer_table_info_config.rs:8]: we had better perf with lower batch and thread count?

Comment (bowenyang007) [reply config/src/config/indexer_table_info_config.rs:8]: More thread lol. I'd keep these as is and let @larry-aptos add the dynamic optimization we applied to that side. We also need a whole lot of testing.

File: api/types/src/convert.rs (hunk 1 above)
Comment (bowenyang007) [reply api/types/src/convert.rs:444]: But what if the flag is turned off but the db is there? I think that it should all still be flag driven?

Comment (jillxuu) [reply api/types/src/convert.rs:444]: it's flag driven but its completely driven by `indexer_table_info` config's `enabled` flag now. thats what i mean when im referring to its separation from aptosdb.

**Review by bowenyang007 — APPROVED** [review]

File: ecosystem/indexer-grpc/indexer-grpc-table-info/Cargo.toml [hunk 3]
```diff
@@ -18,13 +18,16 @@ base64 = { workspace = true }
 bytes = { workspace = true }
 chrono = { workspace = true }
 fail = { workspace = true }
+flate2 = { workspace = true }
```
Comment (larry-aptos) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/Cargo.toml:21]: i am a fan of lz4 now haha

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use flate2::{read::GzDecoder, write::GzEncoder, Compression};
+use std::{
```
Comment (larry-aptos) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs:5]: note: std operations can block the main thread like `std::thread::sleep` if used with tokio tasks. 

should be fine with std threads.

Comment (jillxuu) [reply ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs:5]: yep totally aware of that, in the next pr, creating a standalone spawn blocking thread to separately handle file operation and gcs upload, so that it won't block other concurrent async threads from running.

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs:5]: oh my point is, you can use tokio::fs/tokio::io if IO operations needed in async context. 

spawn_blocking is for CPU intensive operations.

**Review by larry-aptos — APPROVED** [review]

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+/// Creates a tar.lz4 archive from the db snapshot directory
+pub fn create_tar_lz4(
+    dir_path: PathBuf,
+    backup_file_name: &str,
+) -> Result<(PathBuf, String), anyhow::Error> {
+    let tar_lz4_file_name = format!("{}.tar.lz4", backup_file_name);
+    let tar_lz4_file_path = dir_path.join(&tar_lz4_file_name);
+
+    // Create a TAR file in memory
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs:61]: That's a lot of data in memory? Can't we stream to the disk?

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    fs::rename(&temp_file_path, target_path)?;
+    Ok(())
+}
+
+/// Unpack a tar.lz4 archive to a specified directory
+pub fn unpack_tar_lz4(lz4_file_path: &PathBuf, target_dir_path: &PathBuf) -> anyhow::Result<()> {
+    // Read compressed data from LZ4 file
+    let mut lz4_file = File::open(lz4_file_path)?;
+    let mut compressed_data = Vec::new();
+    lz4_file.read_to_end(&mut compressed_data)?;
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs:95]: ditto: memory concern

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        self.set_metadata_epoch(epoch);
+
+        // rocksdb will create a checkpoint to take a snapshot of full db and then save it to snapshot_path
+        indexer_async_v2
+            .create_checkpoint(&snapshot_path)
+            .context(format!("DB checkpoint failed at epoch {}", epoch))?;
+
+        // create a lz4 compressed tar file by compressing a folder into a single file
+        let (tar_file, _tar_file_name) = create_tar_lz4(snapshot_path.clone(), &epoch.to_string())?;
+        let buffer = std::fs::read(&tar_file).context("Failed to read lz4 compressed tar file")?;
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs:186]: memory concern

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+        let filename = generate_blob_name(epoch);
+
+        match self
+            .gcs_client
+            .upload_object(
+                &UploadObjectRequest {
+                    bucket: self.bucket_name.clone(),
+                    ..Default::default()
+                },
+                buffer.clone(),
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs:197]: why the clone?

**Comment by msmouse** [comment]
Are we comfortable operating the entire tarball in memory?

**Comment by jillxuu** [comment]
> Are we comfortable operating the entire tarball in memory?

new commit addressed the memory concern to either write to disk directly or use bufread /bufwrite. also adding spawn_blocking when processing these sync io and fs operations to not block the async thread in general

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    let gz_encoder = GzEncoder::new(tar_file, Compression::default());
+    let tar_data = BufWriter::new(gz_encoder);
+    let mut tar_builder = Builder::new(tar_data);
+
+    tar_builder.append_dir_all(".", &dir_path)?;
+    tar_builder.into_inner()?;
+
+    std::fs::rename(&temp_tar_file_path, &tar_file_path)?;
+
+    Ok((tar_file_path, tar_file_name))
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/fs_ops.rs:73]: `tar_file_name` not used on all the callsites ever?

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+        indexer_async_v2
+            .create_checkpoint(&snapshot_path)
+            .context(format!("DB checkpoint failed at epoch {}", epoch))?;
+
+        // Spawn blocking a thread to create a gzipped tar file by compressing a folder into a single file
+        // synchronously without blocking the async thread
+        let snapshot_path_closure = snapshot_path.clone();
+        let (tar_file, _tar_file_name) = task::spawn_blocking(move || {
+            create_tar_gz(snapshot_path_closure.clone(), &epoch.to_string())
+                .expect("Failed to create tar.gz file in blocking task")
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs:191]: why panicking within spawn_blocking, instead of "this thread"? (then you don't need the closure?)

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        base_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        assert!(metadata.chain_id == chain_id, "Chain ID mismatch.");
+
+        let epoch = metadata.epoch;
+        let epoch_based_filename = generate_blob_name(epoch);
+
+        match self
+            .gcs_client
+            .download_object(
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs:251]: should stream to file instead of getting it in mem.

File: ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+                        Ok(data) => temp_file.write_all(&data).await?,
+                        Err(e) => return Err(anyhow::Error::new(e)),
+                    }
+                }
+                temp_file.sync_all().await?;
+
+                // Spawn blocking a thread to synchronously unpack gzipped tar file without blocking the async thread
+                task::spawn_blocking(move || {
+                    unpack_tar_gz(&temp_file_path_clone, &db_path)
+                        .expect("Failed to unpack gzipped tar file");
```
Comment (msmouse) [inline ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs:282]: nit: I think it's better to panic the current task than in the blocking task runner thread but I don't know why (maybe you will see `restore_db_snapshot` in the stack trace instead of `closure`) :trollface:

**Review by msmouse — APPROVED** [review]
